src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+                )
+            elif image_padding_embed == "pad_and_partial_pad":
+                pad_embed = self.pad_embed[:, None, None, None, :]
+                all_pad = image_masks == 0
+                partial_pad = torch.logical_and(image_masks < 1, torch.logical_not(all_pad)).to(
+                    dtype=image_features.dtype
+                )
+                all_pad = all_pad.to(dtype=image_features.dtype)
+                image_features = image_features + pad_embed[0] * torch.unsqueeze(all_pad, -1)
+                image_features = image_features + pad_embed[1] * torch.unsqueeze(partial_pad, -1)
+            else:
+                raise ValueError(image_padding_embed)
+
+        image_features = self.image_feature_dropout(image_features)
+        num_patches = self.config.image_num_patches
+        image_features = image_features.reshape(
+            (batch_size, patches) + (num_patches, num_patches) + (-1,),
+        )
+
+        if num_patches % self.config.pooling_height == 1:
+            # Pad so we can still pool 2x2 patches
+            image_features = F.pad(
+                image_features,
+                (0, 0, 0, 1, 0, 1, 0, 0, 0, 0),
+            )
+
+        # image pooling
+        leading_dimension, image_batch_size, patch_height, patch_width, image_embed_dim = image_features.shape
+
+        image_features = image_features.view(
+            leading_dimension,
+            image_batch_size,
+            patch_height // self.config.pooling_height,
+            self.config.pooling_height,
+            patch_width // self.config.pooling_width,
+            self.config.pooling_width,
+            image_embed_dim,
+        )
+        image_features = image_features.permute(0, 1, 2, 4, 3, 5, 6).reshape(
+            -1, self.config.pooling_height * self.config.pooling_width, image_embed_dim
+        )
+
+        if self.config.image_pooling_type == "attention_meanq":
+            queries = image_features.mean(-2, keepdim=True)
+            image_features = self.image_pooling_2d(queries, image_features)[0]
+        elif self.config.image_pooling_type not in {None, "stack"}:
+            queries = image_features[:, :1, :]
+            image_features = self.image_pooling_2d(queries, image_features)[0]
+
+        # Round up in case we need to pad the image features for pooling
+        patch_height = (num_patches + self.config.pooling_height - 1) // self.config.pooling_height
+        patch_width = (num_patches + self.config.pooling_width - 1) // self.config.pooling_width
+
+        image_features = image_features.reshape(batch_size, patches, patch_height * patch_width, -1)
+        image_features = self.multi_modal_projector(image_features)
+        return image_features
+
+
+class MolmoForConditionalGeneration(LlavaForConditionalGeneration):
+    def __init__(self, config: MolmoConfig):
+        super().__init__(config)
+        self.adapter = MolmoAdapterModel._from_config(config.pooling_config)
```

Comment:
A comment should explain why we need `_form_config` here. 

Also why did we go with `adapter` naming if we have a conversion script. IMO we should standardize -> `multi_modal_projector` like Llava

Reply:
The initial idea was to enable separate attention impl dispatch on `PoolingAttention` and that would need a new PreTrainedModel with its config. In the current state, i think we dispatch on eager by default when several configs exist and we want to have different attn on vision/text backbones

Yeah, just discovered this flaw as we never had such models before. I will need to update the attn dispatch logic and I'm thinking to make it more similar to `dtype` in composite configs. But that will come later next year anyway.

So for Molmo we can either keep the adapter as is or remove `Adapter` and have it all as part of the `ConditionalModel`. The second case aligns more with VLMs but has an edge case explained above. Maybe we can disregard the edge case given that not many people seem to be using different attn when loading the model